File: frontend/src/services/BannedIndividualsService.ts
import ApiService from './ApiService';
import { BannedIndividual, DetectionAlert } from '../features/banned-individuals/types/banned-individuals.types';

export interface DetectionStatistics {
    total_banned_individuals: number;
    total_detections: number;
    recent_detections: number;
    detection_rate: number;
    period_days: number;
}

class BannedIndividualsService {
    private endpoint = '/banned-individuals';

    private mapToFrontend(data: any): BannedIndividual {
        return {
            id: data.banned_id,
            firstName: data.first_name,
            lastName: data.last_name,
            dateOfBirth: data.date_of_birth,
            nationality: data.nationality || 'UNKNOWN',
            reason: data.reason_for_ban || data.reason || '',
            banType: data.is_permanent ? 'PERMANENT' : 'TEMPORARY',
            banStartDate: data.ban_start_date,
            banEndDate: data.ban_end_date,
            identificationNumber: data.identification_number || '',
            identificationType: data.identification_type || 'Unknown',
            photoUrl: data.photo_url,
            notes: data.notes || '',
            status: data.is_active ? 'ACTIVE' : 'EXPIRED',
            riskLevel: data.is_permanent ? 'CRITICAL' : 'MEDIUM',
            bannedBy: data.added_by || 'System',
            createdAt: data.created_at || data.ban_start_date,
            updatedAt: data.updated_at || data.ban_start_date,
            detectionCount: data.detection_count || 0,
            lastDetection: data.last_detection,
            facialRecognitionData: data.facial_recognition_data,
            // Source tracking
            source: data.source || data.source_type || 'MANAGER',
            sourceMetadata: data.source_metadata || data.sourceMetadata || {}
        };
    }

    private mapToBackend(data: Partial<BannedIndividual>): any {
        const backend: any = {};
        if (data.firstName) backend.first_name = data.firstName;
        if (data.lastName) backend.last_name = data.lastName;
        if (data.dateOfBirth) backend.date_of_birth = data.dateOfBirth;
        if (data.reason) backend.reason_for_ban = data.reason;
        if (data.banType) backend.is_permanent = data.banType === 'PERMANENT';
        if (data.banEndDate) backend.ban_end_date = data.banEndDate;
        if (data.notes) backend.notes = data.notes;
        if (data.photoUrl) backend.photo_url = data.photoUrl;
        if (data.facialRecognitionData) backend.facial_recognition_data = data.facialRecognitionData;
        return backend;
    }

    async getIndividuals(propertyId?: string, isActive: boolean = true): Promise<BannedIndividual[]> {
        let url = this.endpoint;
        const params: any = {};
        if (propertyId) params.property_id = propertyId;
        params.is_active = isActive;

        const response = await ApiService.get<any[]>(url, { params });
        if (response.success && response.data) {
            return response.data.map((item: any) => this.mapToFrontend(item));
        }
        return [];
    }

    async getIndividualById(id: string): Promise<BannedIndividual | null> {
        const response = await ApiService.get<any>(`${this.endpoint}/${id}`);
        return response.success && response.data ? this.mapToFrontend(response.data) : null;
    }

    async createIndividual(data: Partial<BannedIndividual> & { propertyId?: string }): Promise<BannedIndividual | null> {
        const backendData = this.mapToBackend(data);
        if (data.propertyId) backendData.property_id = data.propertyId;

        const response = await ApiService.post<any>(this.endpoint, backendData);
        return response.success && response.data ? this.mapToFrontend(response.data) : null;
    }

    async updateIndividual(id: string, updates: Partial<BannedIndividual>): Promise<BannedIndividual | null> {
        const backendData = this.mapToBackend(updates);
        const response = await ApiService.put<any>(`${this.endpoint}/${id}`, backendData);
        return response.success && response.data ? this.mapToFrontend(response.data) : null;
    }

    async deleteIndividual(id: string): Promise<boolean> {
        const response = await ApiService.delete(`${this.endpoint}/${id}`);
        return !!response.success;
    }

    async getStats(propertyId?: string, days: number = 30): Promise<DetectionStatistics | null> {
        const params: any = {};
        if (propertyId) params.property_id = propertyId;
        params.days = days;

        const response = await ApiService.get<DetectionStatistics>(`${this.endpoint}/stats`, { params });
        return response.success && response.data ? response.data : null;
    }

    async getDetectionAlerts(propertyId?: string, limit: number = 100): Promise<DetectionAlert[]> {
        const params: any = { limit };
        if (propertyId) params.property_id = propertyId;

        const response = await ApiService.get<any[]>(`${this.endpoint}/detections`, { params });
        if (response.success && response.data) {
            return response.data.map((item: any) => {
                const loc = item.location;
                let locationStr: string;
                if (loc == null) locationStr = 'Unknown';
                else if (typeof loc === 'string') locationStr = loc;
                else if (typeof loc === 'object' && 'lat' in loc && 'lng' in loc && typeof loc.lat === 'number' && typeof loc.lng === 'number') {
                    locationStr = `${loc.lat.toFixed(6)}, ${loc.lng.toFixed(6)}`;
                } else locationStr = 'Unknown';
                return {
                id: item.detection_id || item.id,
                individualId: item.banned_id || item.individual_id,
                individualName: item.individual_name || `${item.first_name || ''} ${item.last_name || ''}`.trim(),
                location: locationStr,
                timestamp: item.timestamp || item.detected_at || new Date().toISOString(),
                confidence: item.confidence || 0,
                status: item.status || 'ACTIVE',
                responseTime: item.response_time || 0,
                actionTaken: item.action_taken || 'Detection logged',
                notes: item.notes
                };
            });
        }
        return [];
    }

    async markDetectionAsFalsePositive(detectionId: string, notes?: string): Promise<boolean> {
        const response = await ApiService.put(`${this.endpoint}/detections/${detectionId}/false-positive`, { notes });
        return !!response.success;
    }

    async acknowledgeDetection(detectionId: string, actionTaken: string, notes?: string): Promise<boolean> {
        const response = await ApiService.put(`${this.endpoint}/detections/${detectionId}/acknowledge`, {
            action_taken: actionTaken,
            notes
        });
        return !!response.success;
    }

    async getDetectionFootage(detectionId: string): Promise<string | null> {
        const response = await ApiService.get<{ video_url: string }>(`${this.endpoint}/detections/${detectionId}/footage`);
        return response.success && response.data ? response.data.video_url : null;
    }

    async uploadPhoto(individualId: string, file: File): Promise<{ photoUrl: string } | null> {
        const formData = new FormData();
        formData.append('photo', file);
        formData.append('individual_id', individualId);

        const response = await ApiService.post<{ photo_url: string }>(
            `${this.endpoint}/${individualId}/photo`,
            formData,
            {
                headers: {
                    'Content-Type': 'multipart/form-data'
                }
            }
        );

        if (response.success && response.data) {
            return { photoUrl: response.data.photo_url };
        }
        return null;
    }

    async bulkImport(csvData: string): Promise<{ success: number; failed: number; errors: Array<{ row: number; error: string }> }> {
        interface BulkImportResponse {
            success: number;
            failed: number;
            errors: Array<{ row: number; error: string }>;
        }
        const response = await ApiService.post<BulkImportResponse>(
            `${this.endpoint}/bulk-import`,
            { csv_data: csvData }
        );

        if (response.success && response.data) {
            return response.data;
        }
        return { success: 0, failed: 0, errors: [] };
    }

    async getSettings(propertyId?: string): Promise<any> {
        const params: any = {};
        if (propertyId) params.property_id = propertyId;

        const response = await ApiService.get<any>(`${this.endpoint}/settings`, { params });
        return response.success && response.data ? response.data : null;
    }

    async logAuditEntry(data: {
        actor: string;
        action: string;
        status: 'success' | 'failure' | 'info';
        target?: string;
        reason?: string;
        source?: string;
        metadata?: Record<string, any>;
    }): Promise<boolean> {
        try {
            const response = await ApiService.post(`${this.endpoint}/audit`, data);
            return response.success === true;
        } catch {
            return false;
        }
    }

    async updateSettings(settings: any, propertyId?: string): Promise<boolean> {
        const data: any = { ...settings };
        if (propertyId) data.property_id = propertyId;

        const response = await ApiService.put(`${this.endpoint}/settings`, data);
        return !!response.success;
    }

    async getAnalytics(params: { startDate?: string; endDate?: string; propertyId?: string }): Promise<any> {
        const response = await ApiService.get<any>(`${this.endpoint}/analytics`, { params });
        return response.success && response.data ? response.data : null;
    }
}

export const bannedIndividualsService = new BannedIndividualsService();

File: backend/services/banned_individuals_service.py
"""
Banned Individuals Service
Handles banned individuals management, facial recognition, and detection alerts.
"""

from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
import uuid
import json
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_
from models import BannedIndividual, User, Property
from schemas import BannedIndividualCreate, BannedIndividualResponse
import logging

logger = logging.getLogger(__name__)

class BannedIndividualsService:
    """Service for managing banned individuals and facial recognition."""
    
    def __init__(self, db: Session):
        self.db = db
    
    def add_banned_individual(
        self,
        property_id: str,
        first_name: str,
        last_name: str,
        reason_for_ban: str,
        added_by: str,
        date_of_birth: Optional[datetime] = None,
        photo_url: Optional[str] = None,
        facial_recognition_data: Optional[Dict[str, Any]] = None,
        ban_end_date: Optional[datetime] = None,
        is_permanent: bool = False,
        notes: Optional[str] = None
    ) -> Dict[str, Any]:
        """Add a new banned individual to the database."""
        try:
            banned_individual = BannedIndividual(
                banned_id=str(uuid.uuid4()),
                property_id=property_id,
                first_name=first_name,
                last_name=last_name,
                date_of_birth=date_of_birth,
                photo_url=photo_url,
                facial_recognition_data=facial_recognition_data,
                reason_for_ban=reason_for_ban,
                ban_end_date=ban_end_date,
                is_permanent=is_permanent,
                added_by=added_by,
                notes=notes,
                is_active=True
            )
            
            self.db.add(banned_individual)
            self.db.commit()
            self.db.refresh(banned_individual)
            
            logger.info(f"Added banned individual: {first_name} {last_name}")
            
            return {
                "banned_id": banned_individual.banned_id,
                "name": f"{first_name} {last_name}",
                "status": "active",
                "message": "Banned individual added successfully"
            }
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error adding banned individual: {str(e)}")
            raise e
    
    def get_banned_individuals(
        self,
        property_id: Optional[str] = None,
        is_active: Optional[bool] = None,
        limit: int = 100,
        offset: int = 0
    ) -> List[Dict[str, Any]]:
        """Get list of banned individuals with optional filters."""
        try:
            query = self.db.query(BannedIndividual)
            
            if property_id:
                query = query.filter(BannedIndividual.property_id == property_id)
            
            if is_active is not None:
                query = query.filter(BannedIndividual.is_active == is_active)
            
            banned_individuals = query.offset(offset).limit(limit).all()
            
            result = []
            for individual in banned_individuals:
                result.append({
                    "banned_id": individual.banned_id,
                    "first_name": individual.first_name,
                    "last_name": individual.last_name,
                    "date_of_birth": individual.date_of_birth.isoformat() if individual.date_of_birth else None,
                    "photo_url": individual.photo_url,
                    "reason_for_ban": individual.reason_for_ban,
                    "ban_start_date": individual.ban_start_date.isoformat(),
                    "ban_end_date": individual.ban_end_date.isoformat() if individual.ban_end_date else None,
                    "is_permanent": individual.is_permanent,
                    "added_by": individual.added_by,
                    "detection_count": individual.detection_count,
                    "last_detection": individual.last_detection.isoformat() if individual.last_detection else None,
                    "notes": individual.notes,
                    "is_active": individual.is_active,
                    "created_at": individual.ban_start_date.isoformat()
                })
            
            return result
            
        except Exception as e:
            logger.error(f"Error getting banned individuals: {str(e)}")
            raise e
    
    def get_banned_individual(self, banned_id: str) -> Optional[Dict[str, Any]]:
        """Get a specific banned individual by ID."""
        try:
            individual = self.db.query(BannedIndividual).filter(
                BannedIndividual.banned_id == banned_id
            ).first()
            
            if not individual:
                return None
            
            return {
                "banned_id": individual.banned_id,
                "first_name": individual.first_name,
                "last_name": individual.last_name,
                "date_of_birth": individual.date_of_birth.isoformat() if individual.date_of_birth else None,
                "photo_url": individual.photo_url,
                "facial_recognition_data": individual.facial_recognition_data,
                "reason_for_ban": individual.reason_for_ban,
                "ban_start_date": individual.ban_start_date.isoformat(),
                "ban_end_date": individual.ban_end_date.isoformat() if individual.ban_end_date else None,
                "is_permanent": individual.is_permanent,
                "added_by": individual.added_by,
                "detection_count": individual.detection_count,
                "last_detection": individual.last_detection.isoformat() if individual.last_detection else None,
                "notes": individual.notes,
                "is_active": individual.is_active
            }
            
        except Exception as e:
            logger.error(f"Error getting banned individual {banned_id}: {str(e)}")
            raise e
    
    def update_banned_individual(
        self,
        banned_id: str,
        **updates
    ) -> Dict[str, Any]:
        """Update a banned individual's information."""
        try:
            individual = self.db.query(BannedIndividual).filter(
                BannedIndividual.banned_id == banned_id
            ).first()
            
            if not individual:
                return {"error": "Banned individual not found"}
            
            # Update allowed fields
            allowed_fields = [
                'first_name', 'last_name', 'date_of_birth', 'photo_url',
                'facial_recognition_data', 'reason_for_ban', 'ban_end_date',
                'is_permanent', 'notes', 'is_active'
            ]
            
            for field, value in updates.items():
                if field in allowed_fields and hasattr(individual, field):
                    setattr(individual, field, value)
            
            self.db.commit()
            self.db.refresh(individual)
            
            logger.info(f"Updated banned individual: {banned_id}")
            
            return {
                "banned_id": individual.banned_id,
                "message": "Banned individual updated successfully"
            }
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error updating banned individual {banned_id}: {str(e)}")
            raise e
    
    def delete_banned_individual(self, banned_id: str) -> Dict[str, Any]:
        """Soft delete a banned individual (set is_active to False)."""
        try:
            individual = self.db.query(BannedIndividual).filter(
                BannedIndividual.banned_id == banned_id
            ).first()
            
            if not individual:
                return {"error": "Banned individual not found"}
            
            individual.is_active = False
            self.db.commit()
            
            logger.info(f"Deactivated banned individual: {banned_id}")
            
            return {
                "banned_id": banned_id,
                "message": "Banned individual deactivated successfully"
            }
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error deleting banned individual {banned_id}: {str(e)}")
            raise e
    
    def check_individual(
        self,
        name: Optional[str] = None,
        photo_data: Optional[str] = None
    ) -> Dict[str, Any]:
        """Check if an individual is banned using name or photo."""
        try:
            query = self.db.query(BannedIndividual).filter(
                BannedIndividual.is_active == True
            )
            
            if name:
                # Simple name matching (in real implementation, use fuzzy matching)
                name_parts = name.lower().split()
                for part in name_parts:
                    query = query.filter(
                        or_(
                            BannedIndividual.first_name.ilike(f"%{part}%"),
                            BannedIndividual.last_name.ilike(f"%{part}%")
                        )
                    )
            
            matches = query.all()
            
            if matches:
                return {
                    "is_banned": True,
                    "matches": len(matches),
                    "individuals": [
                        {
                            "banned_id": match.banned_id,
                            "name": f"{match.first_name} {match.last_name}",
                            "reason": match.reason_for_ban,
                            "risk_level": "HIGH" if match.is_permanent else "MEDIUM"
                        }
                        for match in matches
                    ]
                }
            else:
                return {
                    "is_banned": False,
                    "matches": 0,
                    "individuals": []
                }
                
        except Exception as e:
            logger.error(f"Error checking individual: {str(e)}")
            raise e
    
    def facial_recognition_check(self, photo_data: str) -> Dict[str, Any]:
        """Check if a photo matches any banned individuals using facial recognition."""
        try:
            # In a real implementation, this would use actual facial recognition
            # For now, we'll simulate the process
            
            # Get all active banned individuals with facial recognition data
            individuals = self.db.query(BannedIndividual).filter(
                and_(
                    BannedIndividual.is_active == True,
                    BannedIndividual.facial_recognition_data.isnot(None)
                )
            ).all()
            
            # Simulate facial recognition matching
            # In reality, this would use a facial recognition API
            matches = []
            for individual in individuals:
                # Simulate confidence score
                confidence = 85.0 + (hash(photo_data) % 15)  # Random confidence between 85-100
                
                if confidence > 90.0:  # Threshold for match
                    matches.append({
                        "banned_id": individual.banned_id,
                        "name": f"{individual.first_name} {individual.last_name}",
                        "confidence": confidence,
                        "reason": individual.reason_for_ban
                    })
            
            return {
                "match": len(matches) > 0,
                "confidence": max([m["confidence"] for m in matches]) if matches else 0,
                "matches": matches
            }
            
        except Exception as e:
            logger.error(f"Error in facial recognition check: {str(e)}")
            raise e
    
    def record_detection(
        self,
        banned_id: str,
        location: str,
        confidence: float,
        notes: Optional[str] = None
    ) -> Dict[str, Any]:
        """Record a detection of a banned individual."""
        try:
            individual = self.db.query(BannedIndividual).filter(
                BannedIndividual.banned_id == banned_id
            ).first()
            
            if not individual:
                return {"error": "Banned individual not found"}
            
            # Update detection count and last detection
            individual.detection_count += 1
            individual.last_detection = datetime.utcnow()
            
            self.db.commit()
            
            logger.info(f"Recorded detection for banned individual: {banned_id}")
            
            return {
                "banned_id": banned_id,
                "detection_count": individual.detection_count,
                "last_detection": individual.last_detection.isoformat(),
                "message": "Detection recorded successfully"
            }
            
        except Exception as e:
            self.db.rollback()
            logger.error(f"Error recording detection: {str(e)}")
            raise e
    
    def get_detection_statistics(
        self,
        property_id: Optional[str] = None,
        days: int = 30
    ) -> Dict[str, Any]:
        """Get detection statistics for the specified period."""
        try:
            start_date = datetime.utcnow() - timedelta(days=days)
            
            query = self.db.query(BannedIndividual).filter(
                BannedIndividual.is_active == True
            )
            
            if property_id:
                query = query.filter(BannedIndividual.property_id == property_id)
            
            individuals = query.all()
            
            total_detections = sum(individual.detection_count for individual in individuals)
            active_individuals = len(individuals)
            
            # Calculate recent detections (last 7 days)
            recent_detections = 0
            for individual in individuals:
                if individual.last_detection and individual.last_detection >= start_date:
                    recent_detections += 1
            
            return {
                "total_banned_individuals": active_individuals,
                "total_detections": total_detections,
                "recent_detections": recent_detections,
                "detection_rate": total_detections / max(active_individuals, 1),
                "period_days": days
            }
            
        except Exception as e:
            logger.error(f"Error getting detection statistics: {str(e)}")
            raise e
    
    def get_facial_recognition_status(self) -> Dict[str, Any]:
        """Get facial recognition system status and statistics."""
        try:
            # Get all individuals with facial recognition data
            individuals_with_fr = self.db.query(BannedIndividual).filter(
                and_(
                    BannedIndividual.is_active == True,
                    BannedIndividual.facial_recognition_data.isnot(None)
                )
            ).count()
            
            total_active = self.db.query(BannedIndividual).filter(
                BannedIndividual.is_active == True
            ).count()
            
            # Calculate training status
            training_percentage = (individuals_with_fr / max(total_active, 1)) * 100
            
            if training_percentage >= 90:
                training_status = "TRAINED"
            elif training_percentage >= 50:
                training_status = "TRAINING"
            else:
                training_status = "NEEDS_TRAINING"
            
            return {
                "training_status": training_status,
                "training_percentage": round(training_percentage, 1),
                "total_faces": individuals_with_fr,
                "active_individuals": total_active,
                "accuracy": 96.8,  # Simulated accuracy
                "last_training": datetime.utcnow().isoformat()
            }
            
        except Exception as e:
            logger.error(f"Error getting facial recognition status: {str(e)}")
            raise e
    
    def get_detection_alerts(self, property_id: Optional[str] = None, limit: int = 100) -> List[Dict[str, Any]]:
        """Get detection alerts"""
        # Placeholder - in production this would query a detections table
        return []
    
    def mark_detection_false_positive(self, detection_id: str, notes: Optional[str] = None) -> bool:
        """Mark a detection as false positive"""
        # Placeholder - in production this would update the detection status
        return True
    
    def acknowledge_detection(self, detection_id: str, action_taken: str, notes: Optional[str] = None) -> bool:
        """Acknowledge a detection"""
        # Placeholder - in production this would update the detection
        return True
    
    def get_detection_footage(self, detection_id: str) -> Optional[str]:
        """Get video footage URL for a detection"""
        # Placeholder - in production this would return actual footage URL
        return None
    
    def bulk_import(self, csv_data: str) -> Dict[str, Any]:
        """Bulk import banned individuals from CSV"""
        # Placeholder - in production this would parse CSV and import
        return {"success": 0, "failed": 0, "errors": []}
    
    def get_settings(self, property_id: Optional[str] = None) -> Dict[str, Any]:
        """Get banned individuals settings"""
        # Placeholder - in production this would fetch from settings table
        return {
            "auto_share_permanent_bans": True,
            "facial_recognition_alerts": True,
            "strict_id_verification": False,
            "audit_logs_retention": "180",
            "biometric_data_purge": "30",
            "confidence_threshold": 85,
            "retention_days": 180
        }
    
    def update_settings(self, settings: Dict[str, Any]) -> bool:
        """Update banned individuals settings"""
        # Placeholder - in production this would save to settings table
        return True
    
    def update_fr_config(self, confidence_threshold: int, retention_days: int) -> bool:
        """Update facial recognition configuration"""
        # Placeholder - in production this would update config
        return True
    
    def trigger_training(self) -> str:
        """Trigger facial recognition model training"""
        # Placeholder - in production this would start training job
        return str(uuid.uuid4())
    
    def get_analytics(self, start_date: Optional[str] = None, end_date: Optional[str] = None, property_id: Optional[str] = None) -> Dict[str, Any]:
        """Get analytics data"""
        # Placeholder - in production this would calculate real analytics
        return {
            "detection_frequency": "2.3",
            "security_effectiveness": "96.8",
            "monthly_trends": []
        }
